src/psyclone/domain/lfric/lfric_types.py

```diff
@@ -0,0 +1,451 @@
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author J. Henrichs, Bureau of Meteorology
+
+'''This module contains a singleton class that manages LFRic types. '''
+
+
+from collections import namedtuple
+
+from psyclone.domain.lfric import LFRicConstants
+from psyclone.psyir.nodes import Literal
+from psyclone.psyir.symbols import (ArrayType, ContainerSymbol, DataSymbol,
+                                    ImportInterface, INTEGER_TYPE, ScalarType)
+
+
+class LFRicTypes:
+    '''This class implements a singleton that manages LFRic types.
+    Using the 'call' interface, you can query the data type for
+    LFRic types, e.g.:
+    lfric_types = LFRicTypes()
+    rvfdds = lfric_types("RealVectorFieldDataDataSymbol")
+
+    '''
+
+    # Class variable to store the singleton instance
+    _instance = None
+
+    # ------------------------------------------------------------------------
+    def __new__(cls):
+        '''Implement a singleton - only one instance will ever be created.
+
+        :returns: the singleton instance of LFRicTypes.
+        :rtype: :py:class:`psyclone.domain.lfric.LFRicTypes`
+
+        '''
+        if LFRicTypes._instance is None:
+            # Return a new instance. The constructor will set _instance
+            # in this case
+            return super().__new__(cls)
+
+        # Return the existing instance, in which case the constructor will
+        # not re-initialise the internal data structures
+        return LFRicTypes._instance
+
+    # ------------------------------------------------------------------------
+    def __init__(self):
+
+        # Test if this is returning the existing instance, if so, skip
+        # initialisation
+        if self == LFRicTypes._instance:
+            return
+
+        # First time __init__ is called, initialise all data structures.
+        LFRicTypes._instance = self
+
+        self._name_to_class = {}
+
+        self._create_modules()
+        self._create_generic_scalars()
+        self._create_lfric_dimension()
+        self._create_specific_scalars()
+        self._create_fields()
+        # Generate LFRic vector-field-data symbols as subclasses of
+        # field-data symbols
+        for intrinsic in ["Real", "Integer", "Logical"]:
+            name = f"{intrinsic}VectorFieldDataDataSymbol"
+            baseclass = self(f"{intrinsic}FieldDataDataSymbol")
+            self._name_to_class[name] = type(name, (baseclass, ), {})
+
+    # ------------------------------------------------------------------------
+    def __call__(self, name):
+        ''':returns: the class of the required type.
+        :rtype:  Class instance
```

Comment:
> It appears to return a `type`, not an instance of that type. 

I did not want to change the existing functionality. For example, the class created by `LFRicTypes` (`psyir` in the past) is passed to the symbol table to create a symbol if required, e.g. (`kernel_interface.py`):
```
        undf_symbol = self._symbol_table.find_or_create_tag(
            f"undf_{fs_name}",
            symbol_type=lfric_types("NumberOfUniqueDofsDataSymbol"),   <-- needs a class here!
            fs=fs_name, interface=self._read_access)
```

> That said, it might be more useful if it did as there are a lot of parentheses when this is used (e.g. in lfric_alg.py)?

@arporter , I had a look, and I think I've found a way to simplify the usage of `LFRicTypes` a bit, getting rid of one `()` - I got it down to:
```
LFRicTypes("LfricIntegerScalarDataType")
```
instead of:
```
LFRicTypes()("LfricIntegerScalarDataType")
```
That means I need use `__new__` to return a different type (which is supported and explicitly documented in Python). So, `LFRicTypes(XX)` returns whatever the corresponding class (or instance) is, while `LFRicTypes()` will return an instance of LFRicTypes()  (though, I can't see a good reason to actually need that anymore tbh, and might as well just not support this and instead raise an exception if this is tried).

Re string usage: I agree to have a better error message. The advantage of using strings is that a lot of mapping class can be created as class variables, e.g. in `kernel_interface`:
```
class KernelInterface(ArgOrdering):
...
    field_mapping = {
        "integer": "IntegerFieldDataDataSymbol",
        "real": "RealFieldDataDataSymbol",
        "logical": "LogicalFieldDataDataSymbol"}
```

If we would not use strings here and instead rely one way or another on LFRicTypes to exist, this would again mean that we need to instantiate `LFRicConstants` at import time (since this is when these class variables are evaluated) :(
So these initialisations would have to be moved into the constructor (`if not initialised ...`), which would work, but I think your code above is way nicer to read.

The only other choice I see would be to create an enum with all valid dictionary keys, BEFORE the dictionary is created (to avoid that we need to create an instance). And that's a bad code duplication (we need to update this static enum list when something new is added.

Reply:
OK, you've convinced me :-) I did notice that the usage of strings made parameterising the generated classes much simpler. Lets stick with strings and improve the error message. Thank you :-)